sw/airborne/firmwares/rover/guidance/rover_guidance_steering.c

```diff
@@ -0,0 +1,176 @@
+ * any later version.
+ *
+ * paparazzi is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#define AUTOPILOT_CORE_GUIDANCE_C
+
+/** Mandatory dependencies header **/
+#include "firmwares/rover/guidance/rover_guidance_steering.h"
+
+#include "generated/airframe.h"
+
+#include "modules/actuators/actuators_default.h"
+#include "modules/radio_control/radio_control.h"
+#include "autopilot.h"
+#include "navigation.h"
+#include "state.h"
+
+#include "filters/pid.h" // Used for p+i speed controller
+
+#include <math.h>
+#include <stdio.h>
+
+// Guidance control main variables
+rover_ctrl guidance_control;
+
+static struct PID_f rover_pid;
+static float time_step;
+static float last_speed_cmd;
+static uint8_t last_ap_mode;
+
+/** INIT function**/
+void rover_guidance_steering_init(void)
+{
+  guidance_control.cmd.delta = 0.0;
+  guidance_control.cmd.speed = 0.0;
+  guidance_control.throttle  = 0.0;
+
+  last_speed_cmd = 0.0;
+  last_ap_mode   = AP_MODE_KILL;
+
+  guidance_control.speed_error = 0.0;
+  guidance_control.kf = SR_MEASURED_KF;
+  guidance_control.kp = 10;
+  guidance_control.ki = 100;
+
+  init_pid_f(&rover_pid, guidance_control.kp, 0.f, guidance_control.ki, MAX_PPRZ*0.2);
+
+  // Based on autopilot state machine frequency
+  time_step = 1.f/PERIODIC_FREQUENCY;
+}
+
+/** Controler functions **/
+
+/* Steering control (GVF) */
+static void delta_ctrl(void)
+{
+  float delta = 0.0;
+  float omega = guidance_control.gvf_omega; //GVF give us this omega
+
+  // Speed is bounded to avoid GPS noise while driving at small velocity
+  float speed = BoundSpeed(guidance_control.state_speed); 
+
+  if (fabs(omega)>0.0) {
+      delta = DegOfRad(-atanf(omega*DRIVE_SHAFT_DISTANCE/speed));
+    }
+
+  guidance_control.cmd.delta = BoundDelta(delta);
+}
+
+/* Speed control (feed feed forward + propotional + integral controler) (PID)*/
+static void speed_ctrl(void) 
+{
+  // - Looking for setting update
+  if (guidance_control.kp != rover_pid.g[0] || guidance_control.ki != rover_pid.g[2]) {
+    set_gains_pid_f(&rover_pid, guidance_control.kp, 0.f, guidance_control.ki);
+  }
+  if (guidance_control.cmd.speed != last_speed_cmd) {
+    last_speed_cmd = guidance_control.cmd.speed;
+    //reset_pid_f(&rover_pid);
+  }
+
+  // - Updating PID
+  guidance_control.speed_error = (guidance_control.cmd.speed - guidance_control.state_speed);
+  update_pid_f(&rover_pid, guidance_control.speed_error, time_step);
+
+  guidance_control.throttle = BoundThrottle(guidance_control.kf*guidance_control.cmd.speed + get_pid_f(&rover_pid));
+}
+
+
+/** PERIODIC function **/
+void rover_guidance_steering_periodic(void)
```

Comment:
The code of this function is not really in the "spirit" of the autopilot code generation, since the modes here must match the one in the autopilot XML file.
So the different parts of this code should be directly called by the XML file. Just provide some extern functions like `rover_guidance_steering_enter`, `rover_guidance_steering_speed_ctrl`, `rover_guidance_steering_heading_ctrl`, etc.